### DISCUSSION TITLE
Select 等 popover 组件, 支持失焦时不要收起吗?  比如点击控制台时依然展开?

### DISCUSSION THREAD

**Discussion by vaynevayne** [ant-design/ant-design#42863] [Q&A]
不确实是不是 antd  内部在写了 onblur 时收起popover 的逻辑
如果写了 是不是去掉好些? 或者增加个配置也好, 现在的痛点是, 调试非常难

**Comment by yoyo837** [comment]
仅仅为了调试而做成功能? 受控一下是否就可以了?

**Reply by vaynevayne** [reply]
对于非受控模式使用呢? 
而且就算受控模式, 每次调试前还得大改代码,也不合理啊
我猜测 antd  onBlur 时就关了, 貌似应该改为 body 点击时关闭,  不确实使用 useOutsideClick 是不是可以做到? 
现在的开发体验真的难受, 很大期望你们能修改

**Reply by yoyo837** [reply]
我还真就是这么干的, 这几年. 😂

**Comment by zombieJ** [comment]
参考一下这个 #42854

从无障碍角度考虑，输入元素失去焦点后就不会有交互元素展示了。所以才会做成 blur 关闭的。如果没有这方面需求，直接 HOC 一个 Select 然后都用它就行了。

**Comment by vaynevayne** [comment]
https://mui.com/material-ui/react-select/
但是看到mui  是可以点击打开控制台的,他应该也是支持无障碍的

**Comment by vaynevayne** [comment]
我是说真的, 强烈建议修改一下  @afc163  @zombieJ 
比如说github 的这个功能,  关闭弹窗时,popover 也要在, 而之前使用你们的 select, 即使受控模式, 也经常自己关闭了, 就因为 blur 事件


<img width="1060" alt="image" src="https://github.com/ant-design/ant-design/assets/124666577/0595d2cf-d508-4c02-b957-6d8f77339837">

我有在 discuss 里说过这个问题

**Comment by linxianxi** [comment]
在 dom 打断点可以调试的，不会的话搜一下学一下

**Reply by linxianxi** [reply]
在这里面右键 dom 断点可以全部打勾，再打开一个下拉框，慢慢点下一步，就可以调试了
<img width="1286" alt="image" src="https://github.com/ant-design/ant-design/assets/47104575/e02bb3de-c3c4-4d1d-a614-ff78f9c144ee">

**Comment by vaynevayne** [comment]
除了 debug的问题, 还有: 假如内部还有弹窗需要打开时, 这个 popover 会自动关闭,无法维持

**Reply by linxianxi** [reply]
这个试了下，4.x 是可以受控解决的，5.x 有问题，要比较 hack 的方式解决。不太好，我提个 issue

**Reply by vaynevayne** [reply]
我觉得把内部的 blur 事件去掉, 也就没有这么多 bug 要修了

**Reply by linxianxi** [reply]
去掉那怎么关闭？

**Reply by vaynevayne** [reply]
绑定点击事件, 比如给body

**Reply by BoyYangzai** [reply]
> 除了 debug的问题, 还有: 假如内部还有弹窗需要打开时, 这个 popover 会自动关闭,无法维持

@vaynevayne 我理解这用 受控模式 + useClickOutside 不是很好实现么
传入 dom 数组
```typescript
import { MutableRefObject, useCallback, useEffect, useRef } from 'react';

const defaultEvent = 'click';
type RefType = Element | (() => Element | null) | null | undefined;

const useClickOutside = <T extends Element>(
  dom: RefType | RefType[],
  callback: (event: Event) => void,
  eventName: string = defaultEvent,
  listenerOptions?: boolean | AddEventListenerOptions
) => {
  const callbackRef = useRef(callback);
  const targetRefs = useRef<RefType[]>([]);
  const element = useRef<T>();

  useEffect(() => {
    callbackRef.current = callback;
  }, [callback]);

  useEffect(() => {
    targetRefs.current = Array.isArray(dom) ? dom : [dom];
  }, [dom]);

  const eventHandler = useCallback(
    (e: Event) => {
      const isOutside = targetRefs.current.every((item) => {
        const el = typeof item === 'function' ? item() : item || element.current;
        return !el?.contains(e.target as Node);
      });

      if (isOutside) {
        callbackRef.current(e);
      }
    },
    [element.current, targetRefs.current]
  );

  useEffect(() => {
    document.addEventListener(eventName, eventHandler, listenerOptions);
    return () => {
      document.removeEventListener(eventName, eventHandler, listenerOptions);
    };
  }, [eventName, listenerOptions, eventHandler]);

  return element as MutableRefObject<T>;
};

export default useClickOutside;
```

**Reply by vaynevayne** [reply]
这需要 antd select 内部做的, 你不是也说了嘛 5.x 有问题

**Reply by vitalijalbu** [reply]
Would be fine add backdrop to popover, to make it more professional:
Discussion: https://github.com/ant-design/ant-design/discussions/42894
@zombieJ @vaynevayne

**Comment by Aquelys** [comment]
<img width="1245" height="676" alt="Clipboard_Screenshot_1761187348" src="https://github.com/user-attachments/assets/0b5faa37-f3c4-4781-8ce5-fcf520c048f9" />
看看能否解决问题？
